src/librustc_typeck/diagnostics.rs

```diff
@@ -4095,6 +4095,92 @@ assert_eq!(!Question::No, true);
 ```
 "##,
 
+E0604: r##"
+A cast to `char` was attempted on a type other than `u8`.
+
+Erroneous code example:
+
+```compile_fail,E0604
+0u32 as char; // error: only `u8` can be cast as `char`, not `u32`
+```
+
+As the error message indicates, only `u8` can be cast into `char`. Example:
+
+```
+let c = 86u8 as char; // ok!
+assert_eq!(c, 'V');
+```
+"##,
+
+E0605: r##"
+An invalid cast was attempted.
+
+Erroneous code examples:
+
+```compile_fail,E0605
+let x = 0u8;
+x as Vec<u8>; // error: non-primitive cast: `u8` as `std::vec::Vec<u8>`
+
+// Another example
+
+let v = 0 as *const u8; // So here, `v` is a `*const u8`.
+v as &u8; // error: non-primitive cast: `*const u8` as `&u8`
+```
+
+Only primitive types can be cast into each other. Examples:
+
+```
+let x = 0u8;
+x as u32; // ok!
+
+let v = 0 as *const u8;
+v as *const i8; // ok!
+```
+"##,
+
+E0606: r##"
+A cast between non-primitive types was attempted.
```

Comment:
I agree with what you said. I originally proposed the "non-primitive types" wording because before it was "non-scalar types". "non-scalar types were never described in any of the literature I encountered, but a new Rust use would be familiar with "primitive types". But I guess that's not an appropriate term either. I guess the question is what is the "long" definition for what is valid to cast between? Maybe we can find a better term to use here.